Sanitize grist inputs

A line break in the identifier currently makes entry unknown. See https://github.com/ecolabdata/ecospheres/issues/1067.

We should sanitize all inputs to prevent such errors.

[Another case](https://github.com/opendatateam/udata-front-kit-universe/actions/runs/25363327293/job/74369656527#step:6:60):
```
Run uv run universe feed-universe configs/accessibilite-prod.yaml --fail-on-errors 2>&1 | tee logs/accessibilite-prod.log
API for https://www.data.gouv.fr/ ready.
Starting at Tue May  5 07:38:20 2026.
Processing universe topic 'univers-accessibilite'.
Done at Tue May  5 07:38:20 2026 in 0 seconds.
Traceback (most recent call last):
  ...
  File "/home/runner/work/udata-front-kit-universe/udata-front-kit-universe/universe/grist.py", line 50, in _make_entry
    object_class=DatagouvObject.class_from_name(record["Type"]),
                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^
  File "/home/runner/work/udata-front-kit-universe/udata-front-kit-universe/universe/datagouv.py", line 40, in class_from_name
    raise TypeError(f"{name} is not a DatagouvObject")
TypeError:  is not a DatagouvObject
Error: Process completed with exit code 1.
```

Caused by an extra blank line in grist. We should ignore those, and possibly ignore incomplete lines (with a warning) instead of crashing the job.

--- a/tests/test_grist.py
+++ b/tests/test_grist.py
@@ -1,0 +1,130 @@
+from dataclasses import dataclass
+
+import pytest
+from responses import RequestsMock
+
+from universe.config import Config
+from universe.datagouv import Dataservice, Dataset, Organization, Tag, Topic
+from universe.grist import GristAction, GristApi, GristEntry
+
+
+@pytest.fixture
+def api(config: Config) -> GristApi:
+    return GristApi(config.grist.url, config.grist.table, config.grist.token)
+
+
+@dataclass
+class GristRecord:
+    identifier: str | None
+    type: str | None
+    action: str | None = str(GristAction.INCLURE)
+    category: str | None = None
+
+
+class TestGristEntry:
+    @pytest.mark.parametrize(
+        "inputs, expected",
+        [
+            # identifier
+            (
+                [GristRecord(" abc \n ", "dataset")],
+                [GristEntry(identifier="abc", object_class=Dataset)],
+            ),
+            (
+                [GristRecord("", "dataset")],
+                [],
+            ),
+            (
+                [GristRecord(None, "dataset")],
+                [],
+            ),
+            # type
+            (
+                [GristRecord("abc", "dataset")],
+                [GristEntry(identifier="abc", object_class=Dataset)],
+            ),
+            (
+                [GristRecord("abc", "dataservice")],
+                [GristEntry(identifier="abc", object_class=Dataservice)],
+            ),
+            (
+                [GristRecord("abc", "organization")],
+                [GristEntry(identifier="abc", object_class=Organization)],
+            ),
+            (
+                [GristRecord("abc", "tag")],
+                [GristEntry(identifier="abc", object_class=Tag)],
+            ),
+            (
+                [GristRecord("abc", "topic")],
+                [GristEntry(identifier="abc", object_class=Topic)],
+            ),
+            (
+                [GristRecord("abc", "")],
+                [],
+            ),
+            (
+                [GristRecord("abc", None)],
+                [],
+            ),
+            # action
+            (
+                [GristRecord("abc", "dataset", action="inclure")],
+                [GristEntry(identifier="abc", object_class=Dataset, exclude=False)],
+            ),
+            (
+                [GristRecord("abc", "dataset", action="exclure")],
+                [GristEntry(identifier="abc", object_class=Dataset, exclude=True)],
+            ),
+            (
+                [GristRecord("abc", "dataset", action="")],
+                [],
+            ),
+            (
+                [GristRecord("abc", "dataset", action=None)],
+                [],
+            ),
+            # category
+            (
+                [GristRecord("abc", "dataset", category="foo bar")],
+                [GristEntry(identifier="abc", object_class=Dataset, category="foo bar")],
+            ),
+            (
+                [GristRecord("abc", "dataset", category=" foo\nbar \n ")],
+                [GristEntry(identifier="abc", object_class=Dataset, category="foo\nbar")],
+            ),
+            (
+                [GristRecord("abc", "dataset", category="")],
+                [GristEntry(identifier="abc", object_class=Dataset, category=None)],
+            ),
+            (
+                [GristRecord("abc", "dataset", category=None)],
+                [GristEntry(identifier="abc", object_class=Dataset, category=None)],
+            ),
+        ],
+    )
+    def test_get_entries(
+        self,
+        responses: RequestsMock,
+        api: GristApi,
+        inputs: list[GristRecord],
+        expected: list[GristEntry],
+    ):
+        responses.get(
+            url=api.records_url,
+            json={
+                "records": [
+                    {
+                        "fields": {
+                            "Type": input.type,
+                            "Identifiant": input.identifier,
+                            "Action": input.action,
+                            "Categorie": input.category,
+                        }
+                    }
+                    for input in inputs
+                ]
+            },
+        )
+        entries = api.get_entries()
+        assert entries == expected

--- a/universe/grist.py
+++ b/universe/grist.py
@@ -45,6 +45,7 @@ class GristApi:
         )
         r.raise_for_status()
         records = r.json()["records"]
+        # index entries from start=1 to match Grist line-numbering
         entries = [self._make_entry(rec["fields"], idx) for idx, rec in enumerate(records, start=1)]
         return [entry for entry in uniquify(entries) if entry is not None]
 

--- a/universe/grist.py
+++ b/universe/grist.py
@@ -5,7 +5,7 @@ from dataclasses import dataclass
 from enum import auto, StrEnum
 
 from universe.datagouv import DatagouvObject
-from universe.util import uniquify
+from universe.util import sanitize_string, uniquify
 
 
 class GristAction(StrEnum):
@@ -33,21 +33,33 @@ class GristApi:
         self.table = table
         self.token = token
 
+    @property
+    def records_url(self):
+        return f"{self.base_url}/tables/{self.table}/records"
+
     def get_entries(self) -> Sequence[GristEntry]:
         r = requests.get(
-            f"{self.base_url}/tables/{self.table}/records",
+            self.records_url,
             headers={"Authorization": f"Bearer {self.token}", "Content-Type": "application/json"},
             params={"limit": 0},
         )
         r.raise_for_status()
         records = r.json()["records"]
-        return uniquify(self._make_entry(rec["fields"]) for rec in records)
+        entries = [self._make_entry(rec["fields"], idx) for idx, rec in enumerate(records, start=1)]
+        return [entry for entry in uniquify(entries) if entry is not None]
 
     @staticmethod
-    def _make_entry(record: dict[str, str]) -> GristEntry:
+    def _make_entry(record: dict[str, str], index: int) -> GristEntry | None:
+        identifier = sanitize_string(record["Identifiant"])
+        type = record["Type"]
+        action = record["Action"]
+        category = sanitize_string(record.get("Categorie"))
+        if not (identifier and type and action):
+            print(f"Warning: Invalid grist entry line {index}")
+            return None
         return GristEntry(
-            identifier=record["Identifiant"],
-            object_class=DatagouvObject.class_from_name(record["Type"]),
-            exclude=GristAction(record["Action"]) is GristAction.EXCLURE,
-            category=record.get("Categorie"),
+            identifier=identifier,
+            object_class=DatagouvObject.class_from_name(type),
+            exclude=GristAction(action) is GristAction.EXCLURE,
+            category=category,
         )

--- a/universe/util.py
+++ b/universe/util.py
@@ -50,6 +50,11 @@ def normalize_string(string: str) -> str:
     return unicodedata.normalize("NFKD", string).encode("ascii", "ignore").decode("ascii").lower()
 
 
+def sanitize_string(string: str | None) -> str | None:
+    """Return stripped string when not empty, otherwise None"""
+    return (string.strip() if string else None) or None
+
+
 def uniquify[T](iterable: Iterable[T]) -> Sequence[T]:
     """Return list of unique elements from `iterable`, preserving original order"""
     return list(dict.fromkeys(iterable))

